src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -5369,33 +5369,41 @@ std::tuple<int, std::optional<cls_rgw_obj_chain>> RGWRados::send_chain_to_gc(cls
 void RGWRados::delete_objs_inline(const DoutPrefixProvider *dpp, cls_rgw_obj_chain& chain,
                                   const string& tag, optional_yield y)
 {
-  string last_pool;
-  std::unique_ptr<IoCtx> ctx(new IoCtx);
-  int ret = 0;
-  for (auto liter = chain.objs.begin(); liter != chain.objs.end(); ++liter) {
-    cls_rgw_obj& obj = *liter;
-    if (obj.pool != last_pool) {
-      ctx.reset(new IoCtx);
-      ret = rgw_init_ioctx(dpp, get_rados_handle(), obj.pool, *ctx);
-      if (ret < 0) {
-        last_pool = "";
-        ldpp_dout(dpp, 0) << "ERROR: failed to create ioctx pool=" <<
-        obj.pool << dendl;
-        continue;
-      }
-      last_pool = obj.pool;
-    }
-    ctx->locator_set_key(obj.loc);
-    const string& oid = obj.key.name; /* just stored raw oid there */
-    ldpp_dout(dpp, 5) << "delete_objs_inline: removing " << obj.pool <<
-    ":" << obj.key.name << dendl;
+  if (chain.objs.empty()) {
+    return;
+  }
+
+  // initialize an IoCtx for the first object's pool. RGWObjManifest uses the
+  // same pool for all tail objects
+  auto obj = chain.objs.begin();
+
+  librados::IoCtx ioctx;
+  int ret = rgw_init_ioctx(dpp, get_rados_handle(), obj->pool, ioctx);
+  if (ret < 0) {
+    return;
+  }
+
+  // issue deletions in parallel, up to max_aio at a time
+  auto aio = rgw::make_throttle(cct->_conf->rgw_multi_obj_del_max_aio, y);
```

Comment:
@cbodley : I have a basic question here - when I last worked on GC, the OSD was single threaded, and cls ops execute on an OSD, so how does parallel requests for these cls ops work here?

Reply:
the rados cluster can only process one write op at a time per 'placement group'. each object name gets mapped to a specific pg, which is why we can expect cls operations on the same object to be atomic

but rados is highly scalable. each osd can hold up to `mon_max_pg_per_osd=250` pgs, and it's not uncommon for clusters to have hundreds of osds

gc is deleting many different tail objects here, so it's likely that most of them will map to different pgs which can be processed in parallel

Reply:
delete_objs_inline() takes care of tail objects belonging to a particular rgw object. So does each tail object map to a different rados object and hence get mapped to a different pg?

Reply:
correct, a logical "s3 object" in rgw is composed of a head object and zero or more tail objects in rados

striping object data over 4MB rados objects allows GetObject and PutObject to stream these reads/writes in parallel

Reply:
Thank you for the explanation.

Reply:
thanks for review!